feel-engine/src/main/scala/org/camunda/feel/impl/parser/ExpressionValidator.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
+ * under one or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership. Camunda licenses this file to you under the Apache License,
+ * Version 2.0; you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.feel.impl.parser
+
+import org.camunda.feel.FeelEngine.Failure
+import org.camunda.feel.syntaxtree.{Comparison, _}
+
+class ExpressionValidator(externalFunctionsEnabled: Boolean) {
+
+  def validateExpression(expression: Exp): Option[Failure] =
+    validate(expression).headOption
+
+  private def validate(exp: Exp): List[Failure] = exp match {
+    // validate expression
+    case JavaFunctionInvocation(_, _, _) if !externalFunctionsEnabled =>
+      List(Failure(
+        "External functions are disabled. Use the FunctionProvider SPI (recommended) or enable external function in the configuration."))
+
+    // delegate to inner expression
+    case ConstList(items) => items.flatMap(validate)
```

Comment:
This list seems very verbose and error prone. What happens if you extend the expressions? Will the compiler detect that a case is not covered here? (not sure if the last line here is something like a default case)
I would rather expect something like a visitor pattern here. Not sure what the equivalent is in a FP style

Reply:
Yes, the compiler can see if all cases are covered because `Exp` is a sealed trait. If we introduce a new `Exp` type then it needs to be added also to the validation. 

`case _ => ` is the default case

Yes, this works like a visitor pattern.